### PR TITLE
SOW-24: select LPAs

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
## Changes
- add autocomplete component to Form
- get list of LPAs from API and populate autocomplete with it
- [x] fix unit tests

### CODE REVIEW

File: lib/gds-components/autocomplete/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
+
```
Comment (vferraro-scottlogic) [inline lib/gds-components/autocomplete/index.d.ts:1]: I have taken this type definition from [here](https://github.com/alphagov/accessible-autocomplete/issues/535)

File: lib/gds-components/autocomplete/Autocomplete.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import AccessibleAutocomplete, {
```
Comment (vferraro-scottlogic) [inline lib/gds-components/autocomplete/Autocomplete.tsx:1]: autocomplete is not coming from GDS actually so maybe it shouldn't live in this folder

File: lib/api/index.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          })
+          .catch((error) => {
+            reject(new Error(error));
+          });
+      })
+      .catch((error) => {
+        reject(new Error(error));
+      });
+  });
+};
```
Comment (Ro4052) [inline lib/api/index.ts:50]: We can make this more succinct and it's also worth noting that I don't think the way this is throwing the errors is correct. The `Error` constructor only accepts a `string` parameter, which `error` probably wouldn't be (although the actual type throw is `unknown`. We probably want something along the lines of this:

```suggestion
export const fetchLPAs = async (): Promise<LPAResponse> => {
  try {
    const response = await fetch(
      "https://www.planning.data.gov.uk/entity.json?dataset=local-authority&limit=400"
    );
    return response.json();
  } catch (error) {
    if (error instanceof Error) {
      throw new Error("Failed to fetch LPAs: " + error.message);
    }
    
    throw new Error("Failed to fetch LPAs: " + JSON.stringify(error));
  }
};
```

File: lib/api/index.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+type LPAResponse = {
+  entities: LPAEntity[];
+};
+
+type LPAEntity = {
```
Comment (vferraro-scottlogic) [inline lib/api/index.ts:5]: this is the rest of the data we are getting from PDP about the LPA entities. At the moment we are only using the entities names to populate the autocomplete but at some point we might want to look into using more of this data to populate the header (as pointed out by Eddie)

File: lib/gds-components/autocomplete/Autocomplete.tsx (hunk 2 above)
Comment (Ro4052) [reply lib/gds-components/autocomplete/Autocomplete.tsx:1]: I agree

File: lib/timetable-form/Form.tsx [hunk 5]
```diff
@@ -84,6 +87,16 @@ export const Form = (props: React.HTMLAttributes<HTMLDivElement>) => {
...
 
+  useEffect(() => {
+    fetchLPAs()
+      .then((lpas) => {
+        setLPAs(lpas.entities.map((lpa) => lpa.name));
+      })
+      .catch((error: Error) => {
+        console.error(error.message);
+      });
+  }, []);
```
Comment (Ro4052) [inline lib/timetable-form/Form.tsx:98]: We can also rewrite this without using `then` and `catch` (doing it this way also doesn't seem to bring in the correct types):

```suggestion
  useEffect(() => {
    (async () => {
      try {
        const lpas = await fetchLPAs();
        setLPAs(lpas.entities.map((lpa) => lpa.name));
      } catch (error) {
        console.error(error instanceof Error ? error.message : error);
      }
    })();
  }, []);
```

The function could also be named if you didn't want to keep it anonymous:

```suggestion
  useEffect(() => {
    const loadLpas = async () => {
      try {
        const lpas = await fetchLPAs();
        setLPAs(lpas.entities.map((lpa) => lpa.name));
      } catch (error) {
        console.error(error instanceof Error ? error.message : error);
      }
    };
    loadLpas();
  }, []);
```

File: lib/timetable-form/Form.tsx [hunk 6]
```diff
@@ -84,6 +88,18 @@ export const Form = (props: React.HTMLAttributes<HTMLDivElement>) => {
     reader.readAsText(file);
   }, []);
 
+  useEffect(() => {
+    const loadLpas = async () => {
+      try {
+        const lpas = await fetchLPAs();
+        setLPAs(lpas.entities.map((lpa) => lpa.name));
+      } catch (error) {
+        console.error(error instanceof Error ? error.message : error);
```
Comment (Ro4052) [inline lib/timetable-form/Form.tsx:97]: Not for this PR but we need to start thinking more about our error handling and unhappy paths

**Review by Ro4052 — APPROVED** [review]

**Pull request merged**